### DISCUSSION TITLE
Update single items of a large list

### DISCUSSION THREAD

**Discussion by manelio** [pmndrs/jotai#2128] [Q&A]
Hello.

I am using jotai as the state manager for a large application.

I have a question about how to approach a scenario:

In my application a user can load hundreds of files at a time. By means of an event I receive the progress of each item: 10%, 20%, ...

To manage the list of uploads I use an atom:

```
const allFilesAtom = atom([])
```

The uploads arrive to the list as a response to a user action:

```
const add = (newFiles) => {
  setAllFiles(allFiles => [...allFiles, ...newFiles])
  // ...
}
```

And every time I receive a file status change event I update the whole list of files:

```
const [allFiles, setAllFiles] = useAtom(allFilesAtom)


function onUploadProgress(i, percentage) {
  setAllFiles(items => {
    // return the list of uploads with progress of i updated
  })
}

```

I return a `splitAtom` of `allFilesAtom`, so that the components that render the status of each upload are rendered in isolation and avoid rerenders.

My doubt is: is there any way to update only the element reported by the onUploadProgress event? There can be hundreds of files, and the progress update events are constant.

I don't know if I have explained myself correctly.

Thank you!

**Comment by dai-shi** [comment]
One technique is called atoms-in-atom: https://jotai.org/docs/guides/atoms-in-atom

The idea behind `splitAtom` is the same. Using atoms-in-atom directly might be straightforward.

**Comment by manelio** [comment]
Thanks for the quick response.

But I don't think I have explained myself well:

My doubt is more about setting the value than about using the value.

Suppose allFilesAtom is actually allFilesAtomsAtom, an array of atoms.

My problem remains how, from a generic event handler for all items (`onUploadProgress`), to update the value of a single atom.

This is a very reduced skeleton of the structure:

```
import { atom, useAtom } from 'jotai'
import { splitAtom } from 'jotai/utils'

const allFilesAtom = atom([])

export const useUpload = () => {

  const [allFiles, setAllFiles] = useAtom(allFilesAtom)

  const add = (files) => {
    setAllFiles(allFiles => [...allFiles, ...files])
    
    // here I send `files` to be processed by the uploader
  }

  return [allFiles, add]
}


export const UploadProvider = ({
  children,
}) => {

  useEffect(() => {
    uploader.on('upload-progress', (i, progress) => {
      // HERE is my doubt. how to set the value of the element i of the array of files
    })
  }, []);

  return(
  <React.Fragment>
  {children}
  </React.Fragment>
  )

}
```

**Reply by dai-shi** [reply]
Regardless of using `splitAtom` or not, you should update the minimal values that are changed, immutably. It may comes with some complications for nested structure, but for list, it wouldn't be that complicated.

`setAllFiles(allFiles => [...allFiles, ...files])` looks good as it just appends items.

I'm not sure if I get your point of question, though.

**Comment by manelio** [comment]
I think I haven't explained myself well. I'm going to prepare a CodeSandbox.

But in the meantime, I'm going to try to be clearer:

The only doubt is here:

```
useEffect(() => {
    uploader.on('upload-progress', (i, progress) => {
      // HERE is my doubt. How to set the value of the element i of the array of files
    })
  }, []);
```

This event is called every time there is progress in the upload (for simplicity, let's say it's 100 times for each item).

Let's assume we have a list of 100 uploads.

This means that 'upload-progress' will be called 100x100 = 10,000 times.

With my current approach, each call will be something like:

```
export const UploadProvider = ({
  children,
}) => {


  const [allFiles, setAllFiles] = useAtom(allFilesAtom)

  useEffect(() => {
    uploader.on('upload-progress', (i, progress) => {
      setAllFiles(items => [
        ...items.slice(0, i),
        {...items[i], progress},
        ...items.slice(i+1),
      ])
    })
  }, []);

  return(
  <React.Fragment>
  {children}
  </React.Fragment>
  )

}
```

But it seems excessive to set the state of the entire list of items when what I want is to update a single item of the list:

```
uploader.on('upload-progress', (i, progress) => {
  setValueSomehow(allFiles[i], {...allFiles[i], progress})
})
````

**Reply by dai-shi** [reply]
Yeah, please create a minimal example based on https://csb.jotai.org , the smaller the better.
